server/src/main/java/org/elasticsearch/common/time/EpochTime.java

```diff
@@ -169,10 +177,15 @@ public TemporalAccessor resolve(
                 nanos = secondsAndMillis % 1000 * 1_000_000;
                 // `secondsAndMillis < 0` implies negative timestamp; so `nanos < 0`
                 if (nanosOfMilli != null) {
-                    // aggregate fractional part of the input; subtract b/c `nanos < 0`
-                    nanos -= nanosOfMilli;
+                    if (roundUp) {
+                        // these are not the nanos you think they are; these are "round up nanos" not the fractional part of the input
+                        nanos += nanosOfMilli;
```

Comment:
The nanos are negated because the negative is used as a placeholder to indicate rounding.  However, they don't have to be.  I could rewrite the code to not negative the nanos and subtract them subsequently.  This might be every so slightly more performant but I found the current implementation more readable as you can immediately see that the negate nanos which are invalid were intentionally invalid as a placeholder indicating rounding should occur.   I'm up for opinions here.  I'm happy to replace the logic with a large comment indicating that negative nanos is not a bug but an intentional flag here instead of using a boolean to flag it.

Reply:
I agree.  I'm not sure of an alternative.   Do you have any suggestions?  

The parser has to set some available field to indicate during parsing that this is a value that's missing fractional milliseconds and needs to be rounded.  Utilizing the NANOS_OF_MILLIS field is the most intuitive thing to me given the situation.  I had originally tried piggy backing on the negative sign but that misses an edge case.